ownCloud/Client/Actions/Actions+Extensions/ImportPasteboardAction.swift

```diff
@@ -0,0 +1,125 @@
+	override class var locations : [OCExtensionLocationIdentifier]? { return [.moreFolder, .keyboardShortcut] }
+	override class var keyCommand : String? { return "V" }
+	override class var keyModifierFlags: UIKeyModifierFlags? { return [.command] }
+
+	// MARK: - Extension matching
+	override class func applicablePosition(forContext: ActionContext) -> ActionPosition {
+		let pasteboard = UIPasteboard.general
+		if pasteboard.numberOfItems > 0 {
+			return .afterMiddle
+		}
+
+		return .none
+	}
+
+	// MARK: - Action implementation
+	override func run() {
+		guard context.items.count > 0, let viewController = context.viewController, let core = self.core, let rootItem = context.query?.rootItem, let tabBarController = viewController.tabBarController as? ClientRootViewController else {
+			completed(with: NSError(ocError: .insufficientParameters))
+			return
+		}
+
+		let pasteboard = UIPasteboard.general
+
+		// Determine, if the internal pasteboard is the current item and use it
+		if pasteboard.changeCount == tabBarController.pasteboardChangedCounter {
+			// Internal Pasteboard
+			if let pasteboard = UIPasteboard(name: UIPasteboard.Name(rawValue: "com.owncloud.pasteboard"), create: false) {
+				if let data = pasteboard.data(forPasteboardType: "com.owncloud.uti.ocitem.copy"), let object = NSKeyedUnarchiver.unarchiveObject(with: data) {
+					if let item = object as? OCItem, let name = item.name {
+						core.copy(item, to: rootItem, withName: name, options: nil, resultHandler: { (error, _, _, _) in
+							if error != nil {
+								self.completed(with: NSError(ocError: .internal))
+							}
+						})
+					}
+				} else if let data = pasteboard.data(forPasteboardType: "com.owncloud.uti.ocitem.cut"), let object = NSKeyedUnarchiver.unarchiveObject(with: data) {
+					if let item = object as? OCItem, let name = item.name {
+						core.move(item, to: rootItem, withName: name, options: nil) { (error, _, _, _) in
+							if error != nil {
+								self.completed(with: NSError(ocError: .internal))
+							}
+						}
+					}
+				}
+			}
+		} else {
+			// System-wide Pasteboard
+			for type in pasteboard.types {
+				guard let data = pasteboard.data(forPasteboardType: type) else { return }
+				if let extUTI = UTTypeCopyPreferredTagWithClass(type as CFString, kUTTagClassFilenameExtension)?.takeRetainedValue() {
+					let fileName = type
+					let localURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(fileName).appendingPathExtension(extUTI as String)
+					do {
+						try data.write(to: localURL)
+
+						core.importItemNamed(localURL.lastPathComponent,
+											 at: rootItem,
+											 from: localURL,
+											 isSecurityScoped: false,
+											 options: [
+												OCCoreOption.importByCopying : false,
+												OCCoreOption.automaticConflictResolutionNameStyle : OCCoreDuplicateNameStyle.bracketed.rawValue
+							],
+											 placeholderCompletionHandler: { (error, item) in
+												if error != nil {
```

Comment:
@hosy errors returned by SDK are replaced by internal error. Wouldn't it lead to masking of the error in some cases? Like you paste in some file and there is no quota available but the actual error won't be visible in the app?

Reply:
You are right, I replaced it with the SDK errors